src/elfpy/simulators.py

```diff
@@ -347,18 +362,25 @@ def collect_and_execute_trades(self, last_block_in_sim=False):
         """Get trades from the agent list, execute them, and update states"""
         if not isinstance(self.market, Market):
             raise ValueError("market not defined")
+        number_of_executed_trades = 0
         # TODO: This is a HACK to prevent the initial LPer from rugging other agents.
         # The initial LPer should be able to remove their liquidity and any open shorts can still be closed.
         # But right now, if the LPer removes liquidity while shorts are open,
         # then closing the shorts results in an error (share_reserves == 0).
+        wallet_ids = [key for key in self.agents if key > 0]  # exclude init_lp before shuffling
         if self.config.simulator.shuffle_users:
             if last_block_in_sim:
                 wallet_ids = self.rng.permutation(  # shuffle wallets except init_lp
                     [key for key in self.agents if key > 0]  # exclude init_lp before shuffling
                 )
                 wallet_ids = np.append(wallet_ids, 0)  # add init_lp so that they're always last
-            else:
+            else:  # include init_lp only on the last block, to let it unwind
                 wallet_ids = self.rng.permutation(list(self.agents))
+        else:  # we are in a deterministic mode
+            # reverse the list excluding 0 (init_lp)
+            wallet_ids = [key for key in self.agents if key > 0][::-1]
+            if self.config.simulator.init_lp and last_block_in_sim:  # prepend init_lp to the list
```

Comment:
My understanding is that forcing init_lp to be last was a hack to ensure that the end-of-simulation accounting worked out and the printing didn't break. afaik, there's no hard reason why the init_lp agent has to stay in the simulation until the end, or has to be last in the closing trades.

If this is correct, do you think it's possible to update this PR to remedy the situation and remove the hack? If this is not correct, and the ordering is necessary, can you add a comment explaining why?

Reply:
I guess depends whether we want to treat init_lp as an external force or as just another agent. if closer to the latter, i think we want him to act last, since he "opens the door" on initialising liquidity, *then some things happen*, it makes sense to me we'd also ask him to "shut the door" on the way out. then may also want to check if he lost money, or otherwise treat him like any other agent. otherwise, if it's just an external force we don't care about, he doesn't even need to act at the end. but i think i prefer treating him equally.